### PR TITLE
8006 Extend queryparams API

### PR DESCRIPTION
**Pull request by alexander-fedorenko**
## Description
Implemented:
- "map" parameter processing both from GET and POST requests
Example:
```
map={"version":2,"map":{"center":{"x":-66.93461987500001,"y":38.1755548330467,"crs":"EPSG:4326"},"maxExtent":[-20037508.34,-20037508.34,20037508.34,20037508.34],"projection":"EPSG:900913","units":"m","zoom":4,"mapOptions":{},"layers":[{"id":"mapnik__0","group":"background","source":"osm","name":"mapnik","title":"Open Street Map","type":"osm","visibility":true,"singleTile":false,"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"useForElevation":false,"hidden":false},{"id":"Night2012__1","group":"background","source":"nasagibs","name":"Night2012","provider":"NASAGIBS.ViirsEarthAtNight2012","title":"NASAGIBS Night 2012","type":"tileprovider","visibility":false,"singleTile":false,"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"useForElevation":false,"hidden":false},{"id":"OpenTopoMap__2","group":"background","source":"OpenTopoMap","name":"OpenTopoMap","provider":"OpenTopoMap","title":"OpenTopoMap","type":"tileprovider","visibility":false,"singleTile":false,"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"useForElevation":false,"hidden":false},{"id":"s2cloudless:s2cloudless__3","format":"image/jpeg","group":"background","source":"s2cloudless","name":"s2cloudless:s2cloudless","opacity":1,"title":"Sentinel 2 Cloudless","type":"wms","url":["https://1maps.geo-solutions.it/geoserver/wms","https://2maps.geo-solutions.it/geoserver/wms","https://3maps.geo-solutions.it/geoserver/wms","https://4maps.geo-solutions.it/geoserver/wms","https://5maps.geo-solutions.it/geoserver/wms","https://6maps.geo-solutions.it/geoserver/wms"],"visibility":false,"singleTile":false,"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"useForElevation":false,"hidden":false},{"id":"undefined__4","group":"background","source":"ol","title":"Empty Background","type":"empty","visibility":false,"singleTile":false,"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"useForElevation":false,"hidden":false},{"id":"topp:states__2a8624f0-b1a4-11ec-92e7-85acfb394cc7","format":"image/png","search":{"url":"http://localhost:8082/geoserver/wfs","type":"wfs"},"name":"topp:states","description":"This is some census data on the states.","title":"USA Population","type":"wms","url":"http://localhost:8082/geoserver/wms","bbox":{"crs":"EPSG:4326","bounds":{"minx":"-124.731422","miny":"24.955967","maxx":"-66.969849","maxy":"49.371735"}},"visibility":true,"singleTile":false,"allowedSRS":{"EPSG:2154":true,"EPSG:3785":true,"EPSG:3857":true,"EPSG:4269":true,"EPSG:4326":true,"EPSG:102113":true,"EPSG:900913":true},"dimensions":[],"hideLoading":false,"handleClickOnLayer":false,"catalogURL":null,"useForElevation":false,"hidden":false,"version":"1.3.0","params":{}}],"groups":[{"id":"Default","title":"Default","expanded":true}],"backgrounds":[],"bookmark_search_config":{}},"catalogServices":{"services":{"gs_stable_csw":{"url":"https://gs-stable.geo-solutions.it/geoserver/csw","type":"csw","title":"GeoSolutions GeoServer CSW","autoload":true},"gs_stable_wms":{"url":"http://localhost:8082/geoserver/wms","type":"wms","title":"GeoSolutions GeoServer WMS","autoload":false,"oldService":"gs_stable_wms","domainAliases":[]},"gs_stable_wmts":{"url":"https://gs-stable.geo-solutions.it/geoserver/gwc/service/wmts","type":"wmts","title":"GeoSolutions GeoServer WMTS","autoload":false}},"selectedService":"gs_stable_wms"},"widgetsConfig":{"layouts":{"xxs":[],"md":[]}},"mapInfoConfiguration":{"trigger":"click","infoFormat":"text/html"},"dimensionData":{},"timelineData":{}}
```
NOTE: there is limit for url size, it vary from browser to browser so GET method is not reliable for big exports.
- "featureinfo" request processing by re-using existing workflow of identify plugin (dispatching getFeatureInfo action with provided latlng).
- "featureinfo" request - support list of layers to be passed and used as a filter for data request.
Example:
```
featureinfo={"lat": 38.72, "lng": -95.625,"filterNameList":["layerName", "layerName2"]}
```
- GET and POST parameters are processed with a single handler. GET parameters having higher priority and will be used if both GET- and POST-based data is available. Supported parameters: bbox, center, marker, featureinfo, zoom (only when "center" is passed too), actions.

To be done:

- Update documentation

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**
#8006

**What is the new behavior?**
Map configuration can be passed as a separate query parameter. Supporting reading of parameters both from GET request and from sessionStorage (written with POST request and backend)
Identify tool can be triggered for the point & layers passed with query parameter or POST request

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Comment by alexander-fedorenko** [comment]
There are several ways to test how POST approach works.
1. Use browser built-in support to modify and resend requests.
- Open browser tools, networking tab and then open url /rest/config/setParams
- It will complain on unsupported method. On networking tab modify request to the page, change method to POST, add "Content-type: application/json" to the headers and set body of request with json like
```
{
    "map": {},
    "featureinfo": { lat...},
    "page": "#viewer/openlayers/new"
}
```
- Send request, check that it has status 200 and double click on it to open it in new tab.

2. Use supplementary HTML file to make form post request. I used something like this inside web/client directory:
```html
<html><head><meta charset="UTF-8">
<script>
    // Paste JSON or value between backticks
    window.onload = function(e){
        document.getElementById("page").value = '/#/viewer/openlayers/new';
        document.getElementById("map").value = `

    `;
        document.getElementById("featureinfo").value = `

    `;
        document.getElementById("bbox").value = `

    `;
        document.getElementById("center").value = `
        1,1
    `;
        document.getElementById("zoom").value = `
        4
    `;
        document.getElementById("actions").value = `
        []
    `;
        document.getElementById('submit-form').click();
    }
</script>
</head><body>
<form style="display: none;" id="post-form" action="/rest/config/setParams" method="POST" name="myForm">
    <textarea id="map" name="map"></textarea>
    <textarea id="page" name="page"></textarea>
    <textarea id="featureinfo" name="featureinfo"></textarea>
    <textarea id="bbox" name="bbox"></textarea>
    <textarea id="center" name="center"></textarea>
    <textarea id="zoom" name="zoom"></textarea>
    <textarea id="actions" name="actions"></textarea>
    <input id="submit-form" value="Submit" type="submit">
</form>

</body></html>

```

I'm not quite sure what is the real-world use-case for POST approach though. These two can help test that all works fine.

File: web/client/utils/QueryParamsUtils.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Copyright 2022, GeoSolutions Sas.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export const getRequestLoadValue = (name, state) => {
```
Comment (offtherailz) [inline web/client/utils/QueryParamsUtils.js:12]: Document this function (please add parameters and correct if wrong): 
```suggestion
/**
 * Retrieves parameters from hash "query string" of react router
 * Example: `#/viewer/openlayers/0?center=0,0&zoom=5
 */
export const getRequestLoadValue = (name, state) => {
```

**Comment by MV88** [comment]
> @MV88 Please review when it's convenient. Deploying backend based on this PR will provide POST endpoint too.

@alexander-fedorenko this will be reviewed by @offtherailz

File: web/client/utils/QueryParamsUtils.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            if (query[name].length) {
+                return query[name];
+            }
+            return null;
+        }
+    }
+    return null;
+};
+
+export const postRequestLoadValue = (name, storage = sessionStorage) => {
```
Comment (offtherailz) [inline web/client/utils/QueryParamsUtils.js:28]: Document this function (please add parameters and correct if I'm wrong)
Example: 

```suggestion
/**
 * Retrieves parameters from the `queryParams` entry (JSON) of the passed storage (by default `sessionStorage`). 
 * Example: `#/viewer/openlayers/0?center=0,0&zoom=5
 */
export const postRequestLoadValue = (name, storage = sessionStorage) => {
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
Hi I'd like to test it too. In order to test it, we need to have some example. Let's take this opportunity to document this functionality. 

- [ ] Document this in https://mapstore.readthedocs.io/en/latest/developer-guide/map-query-parameters/#retro-compatibility refactoring with this structure: 
- Query Parameters
> MapStore allows to manipulate the map calling it with some paramers. This allows external application to open a customized viewer generating these parameters extenally. With this functionality you can modifiy for instance the initial position of the map, modify the entire map and even trigger some actions. 

- Passing Parameters to the map. 
    - Get Request (how to... with example)
    > The parameters can be passed in a query-string-like section, after the `#<path>?` of the request. Example. 
    - Post Request (fix a little the existing section with concrete example). 
- Available Parameters (one concrete example for each section).
    - Feature Info
    - Map
    - Center
    - Zoom
    - bbox 
    - Actions 
        - ...here the actions present in the doc. 
- [ ] See the comments inline for jsdoc

**Comment by tdipisa** [comment]
@alexander-fedorenko it would be good if you can provide the documentation right now since is seems @offtherailz need it. Please provide is directly where expected at the end, we will review it later after the testing phase only if necessary.

**Comment by alexander-fedorenko** [comment]
@offtherailz Thank you! I've updated documentation and added examples. Please review when it's convenient.

**Review by offtherailz — CHANGES_REQUESTED** [review]
I simplified the documentation and discovered some issues. 

Testing I discovered that **is not possible to post** `application/json` opening the page from JavaScript, so I removed the JSON examples from the form. (only via ajax a post like this is allowed). 
So even if the service allows it, it is technically not possible to use the JSON format. You said you found a way to do it.
Anyway this can not be done because it is not supported by the browser. So I removed the doc for it. 

Moreover I improved a little the test page. You can find it in the updated documentation. 

I spend too much time understanding this and trying to test it, so I ask you to test the various use case (using the sample html improved) and check if: 
- [ ] check everything works fine on your side (testing with the back-end active)
- [ ] improve and clean up the html test page if needed
- [ ] I think page needs  to have (`../../`) for POST. Example: `../../#viewer/openlayers/config` . Please test and confirm. In the case please document and fix the test page. 
- [ ] Make more concrete examples: e.g. `#/mapstore/openlayers/config?center=...` that are **tested**.

**Comment by alexander-fedorenko** [comment]
I've slightly improved example html to simplify the way default data is passed to the form. With this change it makes sense to keep JSON examples for POST request. What do you think?
- New test html works like a charm.
- Did a small cleanup and slight simplification.
- True, to work correctly for for mapstore located at any sub-directory `../../#viewer/openlayers/config` is needed.
- All the examples in doc are those I tried locally, both for GET and POST versions. There are few examples though that are formal ones, like featureinfo, actions to add layers or get feature info, they won't work without real catalogs, or layers on the map. I've updated GET examples to include complete router path, not only query string.

**Review by offtherailz — COMMENTED** [review]
Excellent, it works very good. I cleaned a little bit more.
Just document filterNameList and it is ok for me.

File: docs/developer-guide/map-query-parameters.md [hunk 3]
```diff
@@ -108,25 +102,26 @@ Here a sample page you can create to test the service:
 ## Available Parameters
 
 ### Feature Info
+Allows to trigger [identify tool](../../user-guide/side-bar/#identify-tool) for the coordinates passed in "lat"/"lng" parameters. Optional parameter "filterNameList"
```
Comment (offtherailz) [inline docs/developer-guide/map-query-parameters.md:105]:
```suggestion

Allows to trigger [identify tool](../../user-guide/side-bar/#identify-tool) for the coordinates passed in "lat"/"lng" parameters. Optional parameter "filterNameList"
```

**Review by offtherailz — APPROVED** [review]

**Comment by offtherailz** [comment]
We can merge when build is ready.

**Comment by alexander-fedorenko** [comment]
@ElenaGallo please test it dev. Thanks!

**Comment by offtherailz** [comment]
@ElenaGallo you can find in MapStore documentation here:
https://mapstore.readthedocs.io/en/latest/developer-guide/map-query-parameters/#post-request

some html to copy in a new `html` file (like the embedded) to test the various parameters. 
See the documentation learn how to compile the parameters.

**Comment by ElenaGallo** [comment]
Test passed, @alexander-fedorenko please backport this to 2021.01.xx. Thanks

**Pull request merged**